test/puppetlabs/puppetdb/amq_migration_test.clj

```diff
@@ -183,3 +186,68 @@
              (fn []
                (is (not @mock-upgrade-fn))
                (is (not (needs-upgrade? config)))))))))))
+
+(deftest corrupt-kahadb-journal
+  (testing "corrupt kahadb journal handling"
+    (testing "corruption should return exception"
+      ;; We are capturing the previous known failure here, just in case in the
+      ;; future ActiveMQ changes behaviour (hopefully fixing this problem) so
+      ;; we can make a decision about weither capturing EOFException and
+      ;; restarting the broker ourselves is still needed.
+      ;;
+      ;; Upstream bug is: https://issues.apache.org/jira/browse/AMQ-4339
+      (let [dir (kitchensink/absolute-path (fs/temp-dir "corrupt-kahadb-handling"))
+            broker-name  "test"]
+        (try
+          ;; Start and stop a broker, then corrupt the journal
+          (let [broker (build-embedded-broker broker-name dir {})]
+            (start-broker! broker)
+            (stop-broker! broker)
+            (spit (fs/file dir "test" "KahaDB" "db-1.log") "asdf"))
+          ;; Upon next open, we should get an EOFException
+          (let [broker (build-embedded-broker broker-name dir {})]
+            (is (thrown? java.io.EOFException (start-broker! broker))))
+          ;; Now lets clean up
+          (finally
+            (fs/delete-dir dir)))))
+    (testing "build-and-start-broker! should ignore the corruption"
+      ;; Current work-around is to restart the broker upon this kind of
+      ;; corruption. This test makes sure this continues to work for the
+      ;; lifetime of this code.
+      (let [dir (kitchensink/absolute-path (fs/temp-dir "ignore-kahadb-corruption"))
+            broker-name "test"]
+        (try
+          ;; Start and stop a broker, then corrupt the journal
+          (let [broker (build-embedded-broker broker-name dir {})]
+            (start-broker! broker)
+            (stop-broker! broker)
+            (spit (fs/file dir "test" "KahaDB" "db-1.log") "asdf"))
+          ;; Now lets use the more resilient build-and-start-broker!
+          (let [broker (build-and-start-broker! broker-name dir {})]
```

Comment:
@rbrw any idea why we don't use build-and-start-broker! in the amq-migration code? Reading this test it seems like we should.